### PR TITLE
configurable mongo analytics collection capacity

### LINKED ISSUE #26: Make the tyk_analytics collection capacity configurable in pump.conf
At present, the `tyk_analytics` collection in MongoDB can grow until it hits the in-memory limit. This occurs because the Pump does not apply a cap on the size of the collection, nor does it apply any indexing.
The capacity applied will vary from each installation to another, based on the amount of RAM installed on the machine. A new feature should be added to make the capacity configurable from within the `pump.conf` file.

### PR DESCRIPTION
**Pull request by MangoMM**
resolves #26

### CODE REVIEW

**Comment by mvdan** [comment]
Could you rebase to get rid of the merge commit?

**Comment by MangoMM** [comment]
I tried to do that already. but then it says merge conflict again.

File: pumps/mongo.go [hunk 1]
```diff
@@ -122,7 +217,7 @@ func (m *MongoPump) connect() {
 			"prefix": mongoPrefix,
 		}).Error("Mongo connection failed:", err)
 
-		time.Sleep(5)
+		time.Sleep(5 * time.Second)
```
Comment (buger) [inline pumps/mongo.go]: nice, so previously if mongo failed, it slammed it with numerous connect attempts.

File: pumps/mongo.go [hunk 2]
```diff
@@ -103,6 +115,89 @@ func (m *MongoPump) Init(config interface{}) error {
...
+	if colCapMaxSizeBytes == 0 {
+		defaultBytes := 5
+		colCapMaxSizeBytes = defaultBytes * GiB
+
+		log.WithFields(logrus.Fields{
+			"prefix": mongoPrefix,
+		}).Infof("-- No max collection size set for %s, defaulting to %d", colName, colCapMaxSizeBytes)
+	}
+
+	err = m.dbSession.DB("").C(colName).With(m.dbSession.Copy()).Create(&mgo.CollectionInfo{Capped: true, MaxBytes: colCapMaxSizeBytes})
```
Comment (buger) [inline pumps/mongo.go]: I would say it is too smart :)
Personally, I would prefer splitting it into 2 lines, creating a separate session object and using it.

File: pumps/mongo.go (hunk 1 above)
Comment (MangoMM) [reply pumps/mongo.go]: It still tries forever which IMP needs to be fixed at some stage, but now it actually waits 5 seconds

Comment (mvdan) [reply pumps/mongo.go]: Wow, this is an error that's far too easy to miss. Will check if `go vet` has anything for it.

Comment (mvdan) [reply pumps/mongo.go]: We have a bunch more of those, FYI:

```
tyk/rpc_analytics_purger.go:37:3: time.Sleep(sleep)
tyk-pump/pumps/elasticsearch.go:137:3: time.Sleep(5)
tyk-pump/pumps/influx.go:72:3: time.Sleep(5)
tyk-pump/pumps/mongo.go:125:3: time.Sleep(5)
tyk-pump/pumps/mongo_aggregate.go:333:3: time.Sleep(5)
tyk-pump/pumps/mongo_selective.go:104:3: time.Sleep(5)
tyk-pump/pumps/statsd.go:63:3: time.Sleep(5)
```

Comment (MangoMM) [reply pumps/mongo.go]: good spot. I will do a separate PR to fix that.

Comment (MangoMM) [reply pumps/mongo.go]: https://github.com/TykTechnologies/tyk-pump/issues/55

File: pumps/mgo_helper_test.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	}
+
+	return colStats
+}
+
+func defaultConf() MongoConf {
+	return MongoConf{
+		CollectionName:             colName,
+		MongoURL:                   dbAddr,
+		MongoUseSSL:                false,
```
Comment (mvdan) [inline pumps/mgo_helper_test.go]: you can skip this line

File: pumps/mgo_helper_test.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+	if err := sess.DB("").C(colName).Insert(bson.M{"foo": "bar"}); err != nil {
+		panic(err)
+	}
+}
+
+func (c *Conn) GetCollectionStats() (colStats bson.M) {
+	sess := c.Session.Copy()
+	defer sess.Close()
+
+	if err := sess.DB("").Run(bson.D{{Name: "collStats", Value: colName}}, &colStats); err != nil {
```
Comment (mvdan) [inline pumps/mgo_helper_test.go]: I'd say put the `bson.D` in a separate line to avoid the really long line

File: pumps/mongo.go [hunk 5]
```diff
@@ -11,11 +11,19 @@ import (
 	"gopkg.in/mgo.v2"
 	"gopkg.in/vmihailenco/msgpack.v2"
 
+	"strconv"
```
Comment (mvdan) [inline pumps/mongo.go]: nit: please put this with the rest of the std imports

File: pumps/mongo.go [hunk 6]
```diff
@@ -81,18 +91,20 @@ func (m *MongoPump) Init(config interface{}) error {
 		log.WithFields(logrus.Fields{
 			"prefix": mongoPrefix,
 		}).Info("-- No max batch size set, defaulting to 10MB")
-		m.dbConf.MaxInsertBatchSizeBytes = tenMB
+		m.dbConf.MaxInsertBatchSizeBytes = int(10 * MiB)
```
Comment (mvdan) [inline pumps/mongo.go]: Is the `int()` here necessary? I'm almost certain it isn't, as `10*MiB` should be an untyped integer constant.

File: pumps/mongo.go [hunk 7]
```diff
@@ -81,18 +91,20 @@ func (m *MongoPump) Init(config interface{}) error {
...
 	}
 
 	if m.dbConf.MaxDocumentSizeBytes == 0 {
 		log.WithFields(logrus.Fields{
 			"prefix": mongoPrefix,
 		}).Info("-- No max document size set, defaulting to 10MB")
-		m.dbConf.MaxDocumentSizeBytes = tenMB
+		m.dbConf.MaxDocumentSizeBytes = int(10 * MiB)
```
Comment (mvdan) [inline pumps/mongo.go]: same as above

File: pumps/mongo.go [hunk 8]
```diff
@@ -103,6 +115,92 @@ func (m *MongoPump) Init(config interface{}) error {
...
+
+	if !colCapEnable {
+		return false
+	}
+
+	exists, err := m.collectionExists(colName)
+	if err != nil {
+		log.WithFields(logrus.Fields{
+			"prefix": mongoPrefix,
+		}).Errorf("Unable to determine if collection (%s) exists. Not capping collection: %s\n", colName, err.Error())
```
Comment (mvdan) [inline pumps/mongo.go]: I'm fairly sure that you can skip the trailing newline (here and elsewhere)

**Pull request merged**